### PR TITLE
#6607 - ckeygen should provide a default for the keyfile

### LINKED ISSUE #6607: ckeygen doesn't provide a default for the key file
|[therve's avatar](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#6607|
|Type|enhancement|
|Created|2013-06-30 12:46:51Z|
|Branch|https://github.com/twisted/twisted/tree/ckeygen-default-file-6607-2|

For example:

```
$ ckeygen -p
Enter file in which the key is (/home/therve/.ssh/id_rsa):
```

It shows a file in parenthesis, but doesn't actually use that as a default value if nothing is specified. It should.

Attachments:

* [6607.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1c7/1c7eb6b64f78e03c1fdd3d4455e126ef6a71e674/90f10d0cebc07b540378dea2d7c623b7239edea8.patch) (1535 bytes) - added by Saurabh on 2013-10-09 03:47:30Z - 
* [6607_2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1c7/1c7eb6b64f78e03c1fdd3d4455e126ef6a71e674/e6fc01ed0871bd67d7731d9188f04cc44261313e.patch) (5052 bytes) - added by Saurabh on 2013-11-07 13:33:47Z - 
* [6607_3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1c7/1c7eb6b64f78e03c1fdd3d4455e126ef6a71e674/1dc21ba24ea76dfc2dc75d68e65c2791d282abb4.patch) (5063 bytes) - added by Saurabh on 2013-11-09 04:52:25Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6607 6607
type__enhancement enhancement
reporter__therve therve
priority__low low
milestone__ 
branch__branches_ckeygen_default_file_6607_2 branches/ckeygen-default-file-6607-2
branch_author__lvh__glyph lvh, glyph
status__new new
resolution__ 
component__conch conch
keywords__easy easy
time__1372596411000000 1372596411000000
changetime__1423033410714556 1423033410714556
version__None None
owner__ioggstream ioggstream
cc__z3p
```
</details>

### PR DESCRIPTION
**Pull request by eevelweezel**
## Scope and purpose

Fixes #6607

ckeygen doesn't provide a default for the key file, even though a default appears in the prompt, https://github.com/twisted/twisted/issues/6607

## Contributor Checklist:

This process applies to *all* pull requests - no matter how small.
Have a look at [our developer documentation](https://docs.twisted.org/en/latest/core/development/dev-process.html) before submitting your Pull Request.

Below is a non-exhaustive list (as a reminder):

* The title of the PR should describe the changes and starts with the associated issue number, like “#1234 Brief description”.
* A release notes news fragment file was create in src/twisted/newsfragments/ (see: [Release notes fragments docs.](https://docs.twisted.org/en/latest/core/development/dev-process.html#release-notes-management))
* The automated tests were updated.
* Once all checks are green, request a review by leaving a comment that contains exactly the string `please review`.
  Our bot will trigger the review process, by applying the pending review label
  and requesting a review from the Twisted dev team.

### CODE REVIEW

**Comment by eevelweezel** [comment]
needs-review

File: src/twisted/conch/test/test_ckeygen.py
Comment (adiroiban) [inline src/twisted/conch/test/test_ckeygen.py]: I don't understand the description of this test.
`FileNotFoundError` might be a private implementation details that this test should not care about.

I guess that a description for this test could be.

```suggestion
        When the provided filename via the CLI options is not found
        `displayPublicKey` raises system exits.
        A detail about the error is sent to the standard error.
```

and maybe the name of the tests should be

```diff
-test_handleBadFileDisplayPublicKey
+test_displayPublicKeyHandleFileNotFound
```

Start the name with the target of the test.

Also, it would be nice for the test to assert the condition error.

Maybe SystemExit is raised due to an internal error, and not due to file not found.

File: src/twisted/conch/scripts/ckeygen.py
Comment (adiroiban) [inline src/twisted/conch/scripts/ckeygen.py]: Do we need it as a public API.

----

Public API or private API, I think that this needs a docstring.


---------

I know that `printFingerprint` is public and without documentation, but I think that for the new code we should try to reduce the public interface, and add documentation.

and mabye also add type hinting?

File: src/twisted/conch/scripts/ckeygen.py
Comment (adiroiban) [inline src/twisted/conch/scripts/ckeygen.py]: this is kind of auto-magic code.

Maybe this will make it easier to read and understand what it does

```suggestion
    options["filename"] = resolveFilename(options)
```

File: src/twisted/conch/scripts/ckeygen.py
Comment (adiroiban) [inline src/twisted/conch/scripts/ckeygen.py]: Do we have tests for this ?

File: src/twisted/conch/test/test_ckeygen.py
Comment (adiroiban) [inline src/twisted/conch/test/test_ckeygen.py]: I am not sure that not being empty is a good enough test...

but I don't know what to suggest.

Can we assume that the path will contain the current username ?

**Comment by adiroiban** [comment]
Thanks for helping with this.

I did just a quick review. Feel free to ignore it and wait for another developer to look over this.

File: src/twisted/conch/scripts/ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/scripts/ckeygen.py]: No, this really doesn't need to be public. Agreed, it could use a docstring. Updated.

File: src/twisted/conch/scripts/ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/scripts/ckeygen.py:329]: The tests that cover this are platform-independent, and should pass whether run on Linux or Windows. It's not going to cover both branches in any given run, as a result.

File: src/twisted/conch/test/test_ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/test/test_ckeygen.py]: If options["filename"] is empty, this function will insert the default path for  id_rsa in .ssh, expanded based on user's home and Windows/Linux.  I was trying to keep this minimal so it wouldn't fail based on OS, but we could possibly assert on both "not empty" and "contains 'id_rsa'."

Comment (eevelweezel) [reply src/twisted/conch/test/test_ckeygen.py]: Updated.

File: src/twisted/conch/test/test_ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/test/test_ckeygen.py]: I've amended the test names and added an assert that the error message contains that specific text.

File: src/twisted/conch/scripts/ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/scripts/ckeygen.py]: This should be covered by existing tests when run on a windows image, it just changes the default to a path appropriate for Windows.

File: src/twisted/conch/scripts/ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/scripts/ckeygen.py]: See above, I've added a docstring.

File: src/twisted/conch/scripts/ckeygen.py
Comment (exarkun) [inline src/twisted/conch/scripts/ckeygen.py]: I suggest that `_getKeyOrDefault` avoids mutating `options` and instead returns the value for `filename`.  Then `printFingerprint` can just use that return value directly rather than fishing around inside the `options` object.

Pure functions are better than functions with side-effects and immutable state is better than mutable state.

Twisted has a _lot_ of mutable state all over so I don't think the current `_getKeyOrDefault` behavior is particularly out of place - but almost all of those places in Twisted that work this way are mistakes that, ideally, would gradually be fixed over time.  We can just make this new API avoid this mistake in the first place. :)

File: src/twisted/conch/test/test_ckeygen.py
Comment (exarkun) [inline src/twisted/conch/test/test_ckeygen.py]: It's better to make `input` an argument to `_getKeyOrDefault`.  Monkey-patching, especially builtins, is handy for dealing with APIs that don't accept all of their inputs as parameters - but since this API is under our control, we can just make it accept all of its inputs as parameters.

File: src/twisted/conch/test/test_ckeygen.py
Comment (exarkun) [inline src/twisted/conch/test/test_ckeygen.py]: `assertIn` produces better failure behavior for this case

File: src/twisted/conch/test/test_ckeygen.py
Comment (adiroiban) [inline src/twisted/conch/test/test_ckeygen.py]: Is the assertion called ?

I don't know why the code coverage is not complaining here.

My suggestion is not to use the context here, and instead make use of the nice feature provided by trial :)

```suggestion
        exception = self.assertRaises(SystemExit, displayPublicKey, options)
        self.assertIn("could not be opened, please specify a file.", e)
```

File: src/twisted/conch/scripts/ckeygen.py
Comment (adiroiban) [reply src/twisted/conch/scripts/ckeygen.py]: I think I have also commented about returning a value, instead of mutating the input arguments.

so I guess that this is required for approving this PR

File: src/twisted/conch/scripts/ckeygen.py
Comment (adiroiban) [inline src/twisted/conch/scripts/ckeygen.py]: -----------

And as commented by JP, this function should also allow dependency injection, so that we don't need to path the input, but pass our own custom input method during the tests.
```suggestion
def _getKeyOrDefault(options, input_collector=None):
    """
    If C{options["filename"]} is None, prompt the user to enter a path
    or attempt to set it to .ssh/id_rsa
    """
    if input_collector is None:
        input_collector = input
```

-------

But also it should return the path, instead of mutating the input argument.

File: src/twisted/conch/scripts/ckeygen.py
Comment (adiroiban) [inline src/twisted/conch/scripts/ckeygen.py]:
```suggestion
            input_collector("Enter file in which the key is (%s): " % filename) or filename
```

File: src/twisted/conch/test/test_ckeygen.py
Comment (adiroiban) [inline src/twisted/conch/test/test_ckeygen.py]: There is a lot of repetition in these tests... but this is just a minor comment.

**Review by adiroiban — DISMISSED** [review]
Thanks for the update.

So I think the main things that need to change

* Don't mutate the input argument
* Allow dependency injection
* Fix the test to have the assertion called

File: src/twisted/conch/scripts/ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/scripts/ckeygen.py]: Updated so that _getKeyOrDefault doesn't mutate options.

File: src/twisted/conch/test/test_ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/test/test_ckeygen.py]: The problem is that prompting the user before calling _getKeyOrDefault means we don't have a suggested/default path to include in the prompt. We can't just use a string, because that will only work for Windows or Linux, not both.

File: src/twisted/conch/scripts/ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/scripts/ckeygen.py]: The problem is that prompting the user before calling _getKeyOrDefault means we don't have a suggested/default path to include in the prompt. We can't just use a string, because that will only work for Windows or Linux, not both.

File: src/twisted/conch/test/test_ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/test/test_ckeygen.py]: It's not possible to inspect the SystemExit outside of a context manager.  The first line succeeds, but the second is stopped by a TypeError.  
`[ERROR]
Traceback (most recent call last):
  File "/home/eevel/twisted/src/twisted/conch/test/test_ckeygen.py", line 662, in test_displayPublicKeyHandleFileNotFound
    self.assertIn("could not be opened, please specify a file.", e)
  File "/home/eevel/twisted/src/twisted/trial/_synctest.py", line 505, in assertIn
    if containee not in container:
builtins.TypeError: argument of type 'SystemExit' is not iterable
`
Based on stdlib unittest, this test should only succeed if the first line raises a SystemExit, and the second fulfills the assertTrue assertion.

File: src/twisted/conch/test/test_ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/test/test_ckeygen.py]: Agreed, updated.

File: src/twisted/conch/scripts/ckeygen.py
Comment (adiroiban) [reply src/twisted/conch/scripts/ckeygen.py]: I don't understand the issue.

Here is the complete implementation, I hope it helps describe how I see this method implemented.

```python
def _getKeyOrDefault(options, input_callback=input):
    """
    If C{options["filename"]} is None, prompt the user to enter a path
    or attempt to set it to .ssh/id_rsa.

    @param input_collector: A callback used to collect user input.
      Default to builting L{input} method.
    """
    filename = options["filename"]
    if not filename:
        filename = os.path.expanduser("~/.ssh/id_rsa")
        if platform.system() == "Windows":
            filename = os.path.expandvars(R"%HOMEPATH %\.ssh\id_rsa")
        filename = input_callback("Enter file in which the key is (%s): " % filename) or filename
    return 
```

File: src/twisted/conch/test/test_ckeygen.py
Comment (adiroiban) [reply src/twisted/conch/test/test_ckeygen.py]: one very raw suggestion , just an idea of how this can be used.


```python
    def test_useDefaultForKey(self):
        """
        L{options} will default to "~/.ssh/id_rsa" if the user doesn't
        specify a key using the CLI argument nor via the interactive text input.
        """
        input_prompts = []

        def rigged_input(prompt):
            """
            An input that records the for which the user alwasys insert
            an empty text.
            """
            input_prompts.append()
            return ""

        options = {"filename": ""}
        filename = _getKeyOrDefault(options, rigged_input)

        # Resolved path is an RSA key inside .ssh dir.
        self.assertTrue(filename.endswith(os.path.join(".ssh", "id_rsa")))
        # The user is prompted once to enter the path, since no path was
        # provided via CLI.
        self.assertEqual(1, len(input_prompts))
        self.assertTrue(input_prompts.startswith("Enter file in which the key is ("))

```

**Comment by adiroiban** [comment]
Thanks for the updates. I left some comments.

Once this is ready for another review, please let me know

File: src/twisted/conch/test/test_ckeygen.py
Comment (adiroiban) [reply src/twisted/conch/test/test_ckeygen.py]: maybe

```
self.assertIn("could not be opened, please specify a file.", e.args[0])
```

Comment (exarkun) [reply src/twisted/conch/test/test_ckeygen.py]: Just to be clear - the reason it "works" inside the `assertRaises` is that it never runs.  The `displayPublicKey` call raises an exception and execution jumps to the `assertRaises` context manager exit handler.

File: src/twisted/conch/test/test_ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/test/test_ckeygen.py]: Updated to add dependency injection.

File: src/twisted/conch/scripts/ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/scripts/ckeygen.py]: Updated.

File: src/twisted/conch/scripts/ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/scripts/ckeygen.py]: Updated.

File: src/twisted/conch/test/test_ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/test/test_ckeygen.py]: Updated.

File: src/twisted/conch/test/test_ckeygen.py
Comment (adiroiban) [inline src/twisted/conch/test/test_ckeygen.py]: Thanks for the updates and followp.


Just a remider. 
I see these tests calling assertion *after* an exception is raised.
In this case, the code after is never called... so the assertion is never raised.

I think this is verry bad habit 

My suggestion is to try TDD.

Write the test first without updating the implementation.
See that the test fails.
Update your implementation.
See the test pass.

If the test pass from the first run, is harder to tell if the test really does its job.

Cheers

Comment (eevelweezel) [reply src/twisted/conch/test/test_ckeygen.py]: Updated to remove the context managers.

**Comment by eevelweezel** [comment]
This should address the requested changes, I see that static-checks are also failing in trunk.

**Comment by eevelweezel** [comment]
please review

**Comment by adiroiban** [comment]
I don't think I will have time soon to check this PR... I expect that it looks good and it only needs a quick check.

You might want to do a review exchange with another Twisted  dev.
Review a PR from another dev and then you should get a review back :)

Good luck

File: src/twisted/conch/scripts/ckeygen.py
Comment (glyph) [inline src/twisted/conch/scripts/ckeygen.py]: Can you put some type annotations on here?  New code should have them, but since this is an old file it gets (incorrectly) grandfathered in.

Comment (glyph) [reply src/twisted/conch/scripts/ckeygen.py]: Another nitpick here: technically according to the Twisted coding standard this should be `inputCollector`.

File: src/twisted/conch/scripts/ckeygen.py
Comment (glyph) [inline src/twisted/conch/scripts/ckeygen.py]: Suggestion to implement these comments (1/2):

```suggestion
from typing import Dict, Optional, Callable
def _getKeyOrDefault(options: Dict[str, Optional[str]], inputCollector: Optional[Callable[[str],str]]=None) -> str:
    """
    If C{options["filename"]} is None, prompt the user to enter a path
    or attempt to set it to .ssh/id_rsa
    """
    if inputCollector is None:
        inputCollector = input
    filename = options["filename"]
    if not filename:
```

File: src/twisted/conch/scripts/ckeygen.py
Comment (glyph) [inline src/twisted/conch/scripts/ckeygen.py]: Second half of the suggestion since github has some weird requirements on suggestion boundaries:

```suggestion
        if platform.system() == "Windows":
            filename = os.path.expandvars(R"%HOMEPATH %\.ssh\id_rsa")
        filename = (
            inputCollector("Enter file in which the key is (%s): " % filename)
            or filename
        )
    return filename
```

File: src/twisted/conch/scripts/ckeygen.py
Comment (glyph) [inline src/twisted/conch/scripts/ckeygen.py]: Should be `inputCollector` here as well, also some type annotations

File: src/twisted/conch/test/test_ckeygen.py
Comment (glyph) [inline src/twisted/conch/test/test_ckeygen.py]: Let's not accidentally turn off type-checking in the tests

```suggestion
    def test_printFingerprintSuffixAppended(self) -> None:
```

File: src/twisted/conch/test/test_ckeygen.py
Comment (glyph) [inline src/twisted/conch/test/test_ckeygen.py]:
```suggestion
    def test_saveKeyFileExists(self) -> None:
```

File: src/twisted/conch/test/test_ckeygen.py
Comment (glyph) [inline src/twisted/conch/test/test_ckeygen.py]:
```suggestion
    def test_useDefaultForKey(self) -> None:
```

File: src/twisted/conch/test/test_ckeygen.py
Comment (glyph) [inline src/twisted/conch/test/test_ckeygen.py]:
```suggestion
    def test_displayPublicKeyHandleFileNotFound(self) -> None:
```

File: src/twisted/conch/test/test_ckeygen.py
Comment (glyph) [inline src/twisted/conch/test/test_ckeygen.py]:
```suggestion
    def test_changePassPhraseHandleFileNotFound(self) -> None:
```

File: src/twisted/conch/test/test_ckeygen.py
Comment (glyph) [inline src/twisted/conch/test/test_ckeygen.py]:
```suggestion
    def test_printFingerprintHandleFileNotFound(self) -> None:
```

File: src/twisted/conch/scripts/ckeygen.py
Comment (glyph) [inline src/twisted/conch/scripts/ckeygen.py]:
```suggestion
        yn = inputCollector("Overwrite (y/n)? ")
```

File: src/twisted/conch/scripts/ckeygen.py
Comment (glyph) [inline src/twisted/conch/scripts/ckeygen.py]: It looks like this logic is now expressed twice, both here and in `_getKeyOrDefault`.  Would you mind refactoring it into one place so it's a bit easier to see, rather than hard-coding `_getKeyOrDefault` to RSA keys only?

File: src/twisted/conch/scripts/ckeygen.py
Comment (glyph) [inline src/twisted/conch/scripts/ckeygen.py]:
```suggestion
    @param inputCollector: Dependency injection for testing.
    @type inputCollector: L{function} or None.
```

File: src/twisted/conch/scripts/ckeygen.py
Comment (glyph) [inline src/twisted/conch/scripts/ckeygen.py]: Huh. Not your bug, but this is such an odd choice to not use the path object here…

```suggestion
    fp = filepath.FilePath(filename)
    fp.setContent(
```

File: src/twisted/conch/scripts/ckeygen.py
Comment (glyph) [inline src/twisted/conch/scripts/ckeygen.py]: Also this magic number is terrible :)

```suggestion
    fp.chmod(0o100600)
```

Comment (glyph) [reply src/twisted/conch/scripts/ckeygen.py]: Also we should file a bug here, there's arguably a umask race in this code, we should not be chmod'ing things after the fact

**Review by glyph — CHANGES_REQUESTED** [review]
This looks like a great change and I'm sorry for nit-picking it to death here, but it would be great to give the type-checker a chance to spot a few more bugs before we land it. Would you mind making a few changes? I tried to put in as many of them as suggestions as I could.

File: src/twisted/conch/scripts/ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/scripts/ckeygen.py]: Updated to add typing.

File: src/twisted/conch/scripts/ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/scripts/ckeygen.py]: Added typing

File: src/twisted/conch/scripts/ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/scripts/ckeygen.py]: Changed to camelCase.

File: src/twisted/conch/test/test_ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/test/test_ckeygen.py]: Added return types for tests.

File: src/twisted/conch/scripts/ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/scripts/ckeygen.py]: Updated to camelCase.

File: src/twisted/conch/scripts/ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/scripts/ckeygen.py]: Refactored and updated tests, "rsa" is now the default param if no key type is supplied.

File: src/twisted/conch/scripts/ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/scripts/ckeygen.py]: Updated

File: src/twisted/conch/scripts/ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/scripts/ckeygen.py]: Definitely makes sense, updated.

File: src/twisted/conch/scripts/ckeygen.py
Comment (eevelweezel) [reply src/twisted/conch/scripts/ckeygen.py]: Updated

File: src/twisted/conch/scripts/ckeygen.py
Comment (adiroiban) [inline src/twisted/conch/scripts/ckeygen.py]: no need to add `@type` here, as long as we have type hinting.

and I think is best not to have `@type` in the docstring to avoid future conflicts.

Comment (eevelweezel) [reply src/twisted/conch/scripts/ckeygen.py]: Updated.

**Comment by eevelweezel** [comment]
please review

**Comment by adiroiban** [comment]
I will try to get this reviewed next week, in exchange for your help with the other PRs :)

But if anyone else wants to review it, I am happy to have this merge with a review from another person.

**Review by glyph — APPROVED** [review]

**Comment by glyph** [comment]
Looks like all the feedback was addressed!

**Pull request merged**
